### PR TITLE
Push notifications

### PR DESCRIPTION
**Pull request by VasylShtenovych**
Closes back-end part of ita-social-projects/Car-Back-End/issues/321

### CODE REVIEW

**Review by Nazarii-Mandiak — APPROVED** [review]

**Review by Vlas2001 — APPROVED** [review]

File: Car.Backend/Car.Domain/Services/Implementation/FirebaseService.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+        public FirebaseService(
+            IWebHostEnvironment webHostEnvironment,
+            IOptions<FirebaseOptions> firebaseOptions)
+        {
+            var keyFilePath = Path.Combine(
+                webHostEnvironment.WebRootPath,
+                firebaseOptions.Value.CredentialsPath!);
+
+            var stream = new FileStream(keyFilePath, FileMode.Open, FileAccess.Read);
```
Comment (serhii-zinkiv) [inline Car.Backend/Car.Domain/Services/Implementation/FirebaseService.cs:31]: Don't we need to use using for any stream for resources disposing of?

File: Car.Backend/Car.Domain/Services/Implementation/FirebaseService.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+            var credential = GoogleCredential.FromStream(stream);
+            credential = credential.CreateScoped("https://www.googleapis.com/auth/firebase.messaging");
+
+            App = FirebaseApp.Create(new AppOptions()
```
Comment (serhii-zinkiv) [inline Car.Backend/Car.Domain/Services/Implementation/FirebaseService.cs:35]: Do we need this property as public?

File: Car.Backend/Car.Domain/Services/Implementation/FirebaseService.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+using System.Threading.Tasks;
+using Car.Domain.Configurations;
+using Car.Domain.Services.Interfaces;
+using FirebaseAdmin;
+using FirebaseAdmin.Messaging;
+using Google.Apis.Auth.OAuth2;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.Extensions.Options;
+
+namespace Car.Domain.Services.Implementation
```
Comment (serhii-zinkiv) [inline Car.Backend/Car.Domain/Services/Implementation/FirebaseService.cs:15]: Refactor it.

File: Car.Backend/Car.Domain/Services/Implementation/PushNotificationService.cs [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+        public async Task<bool> SendNotificationAsync(model.Message message)
+        {
+            var chat = await chatRepository.Query()
+                .Where(chat => chat.Id == message.ChatId)
+                .Include(chat => chat.Journey)
+                    .ThenInclude(jour => (jour != null) ? jour.Participants : null)
+                .FirstOrDefaultAsync();
+
+            if (chat == null || chat.Journey == null)
```
Comment (serhii-zinkiv) [inline Car.Backend/Car.Domain/Services/Implementation/PushNotificationService.cs:58]: Use ?

File: Car.Backend/Car.Domain/Services/Implementation/PushNotificationService.cs [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+            {
+                users.Add(chatOwner);
+            }
+
+            foreach (var user in users)
+            {
+                if (user.FCMToken != null && user.Id != message.Sender?.Id)
+                {
+                    await firebaseService.SendAsync(
+                        CreateNotification((message.Sender != null) ? message.Sender.Name : "User", message.Text, user.FCMToken, "MessagesTabs"));
```
Comment (serhii-zinkiv) [inline Car.Backend/Car.Domain/Services/Implementation/PushNotificationService.cs:76]: create some property for first parameter somewhere above

File: Car.Backend/Car.Domain/Services/Implementation/PushNotificationService.cs [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+                }
+            }
+
+            return true;
+        }
+
+        // internal for testing
+        internal static (string Title, string Message) FormatToMessage(model.User sender, NotificationDto notification)
```
Comment (serhii-zinkiv) [inline Car.Backend/Car.Domain/Services/Implementation/PushNotificationService.cs:84]: If you need some method like this you need to take it out somewhere in the static helper class

**Pull request merged**